merkle-patricia-tree/src/Data/Tree/MerklePatricia.hs

```diff
@@ -98,7 +103,51 @@ instance HashAlgorithm alg => Eq (MerklePatriciaTree alg a) where
 -- | An opaque 'Proof' which can be used to prove membership or atomic
 -- operations such as 'add' or 'delete'
 newtype Proof alg = Proof {unProof :: [([Bit], [(Bit, Digest alg)])]}
-  deriving stock (Show)
+  deriving stock (Eq, Show)
+
+instance HashAlgorithm alg => Serialise (Proof alg) where
+  encode (Proof prf) = encodeListLen (fromIntegral $ length prf) <> mconcat (map encodeItem prf)
+   where
+    encodeItem (bits, digests) =
+      -- NOTE(AB): we need to pad the bits length to 8-bit words multiples, but we want
+      -- to keep track of the original length for deserialisation
+      -- TODO: Find a better encoding for the Proof object itself
+      encodeListLen (fromIntegral $ length bits)
+        <> encodeBytes (unsafeFromBits $ pad bits)
+        <> encodeListLen (fromIntegral $ length digests)
+        <> mconcat (map encodeDigest digests)
```

Comment:
What about using a `map` to encode the tuple such that it's "future-proof", we can easily add elements to it later (I mean, instead of the concatenation of two arrays, make it a map with two fields `00` and `01` for the prefix, and the node's hashes respectively.)

Reply:
Unsure what you mean actually...

Reply:
Well, right now, this is not even valid CBOR actually :sweat_smile:, because it is the concatenation of two independent structures, which is not valid. At the root, the whole binary result must be _structured_. To encode a tuple, you have mainly two options:

- An array of 2 elements which is about what you do here I believe, except that it's missing the surrounding array of length 2. So it'll be something like:
   
    ```hs
    mconcat
        [ encodeListLen 2
        , encodeFirstItem ...
        , encodeSecondItem ...
        ] 
    ```

- A CBOR-map, with explicit fields for each element. Like it's done for transaction's body in Cardano for instance. This is more flexible because it allows to omit some elements if they are not necessary, does not rely on the order and, allows for better documenting the binary formal over all. So here, that'll be something like:

   
    ```hs
    mconcat
        [ encodeMapLen 2
        , encodeWord8 0
        , encodeFirstItem ...
        , encodeWord8 1
        , encodeSecondItem ...
        ] 
    ```

   That makes it easier to then document that items at key `00` refers to X and items at key `01` refers to Y. All in all, if you don't need to _omit_ fields, arrays and maps are kind of equivalent in CBOR, but I find maps a bit more flexible which is always good for binary formats :no_mouth:

Reply:
Right, I have started writing the latter but got confused because at the end, it's still concatenating items and you _know_ the order when serialising. At least, I will fix the array thing but I am not convinced it's actually useful to add the tag encoding for future-proofing without adding a version number which would allow future deserialisers to understand past ones. If that makes sense...

Reply:
I will trust your judgment and just add it though :)

Reply:
Perhaps it's not valid CBOR but I can actually serialise it BTW 😂

Reply:
Yes, the `map` is debatable here since both elements must be always present. So it's merely for self-documenting; yet at least, you need a wrapping array around both :sweat_smile: ...  otherwise, it's missing a bit the point of CBOR which is meant to be "self-describing".